setendings.py

```diff
@@ -0,0 +1,624 @@
+logging.basicConfig(format='', level=logging.DEBUG)
+
+# Helper functions to abstract common tasks. Global for now..
+# these functions are associated to types in the transforms dictionary in the
+# TransformVisitor class. The _transform method invokes the transform function
+# on the nodes.
+# TODO: Separate into Class? .. would need static methods.
+################################################################################
+def _set_end_lineno(node):
+    """Transform the property in-place.
+    """
+    node.end_lineno = node.lineno
+
+def _set_end_lineno(node):
+    """Transform the property in-place.
+    Testing with simple, single-line things for now.
+    Goal is to work on various multi-line nodes, with complicated nesting.
+    'tolineno' is set correctly on everything, by asteroid.
+    """
+    # Guard. 
+    if node.tolineno is None:
+        # TODO: write new code to gracefully handle, if this is ever reached.
+        logging.error("⛔ lineno is None at node: {}".format(node))
+        sys.exit(1)
+    else:
+        node.end_lineno = node.tolineno
+
+def _set_end_col_offset_by_string_old(node):
+    """Set the end_col_offset property by the as_string attribute.
+    Uses the end column location of the last line, instead of the widest line.
+    """
+    
+    # Get the widest line in the source code.
+    lines = [line for line in node.as_string().split('\n')]
+    max_width_line = max(lines, key=len)
+    max_width = len(max_width_line)
+    # Note: node.col_offset could be 'None'
+    node.end_col_offset = (node.col_offset or 0) + max_width
+
+
+
+def _set_end_col_offset_by_string(node):
+    """Set the end_col_offset property by the as_string attribute.
+    Uses the end column location of the last line, instead of the widest line.
+    Need to do a custom postorder traversal to set end_col_offset property of 
+    child nodes first, since it is sometimes unset in the last child.
+    Postorder setting of children results in the assumption: if a node has the
+    propert set, then all of its children are set.
+    """
+
+    # Compare with "None" since value could be "0", which is falsey.
+    # Returning a.s.a.p. reduces the runtime (considerably?).
+    if node.end_lineno is not None and node.end_col_offset is not None:
+        return  # since properties are already set.
+
+    # Doing this first makes it a postorder traversal.
+    for child in node.get_children():  # get_children() returns a generator.
+        _set_end_col_offset_by_string(child)
+
+    node_as_string = str(node)
+    node_name = node_as_string[ : node_as_string.find('(')]
+    # print('node: ', node_name)
+
+    #❕TODO: set up the below to work with postorder traversal, setting all
+    # child properties, calling the different transform functions by node
+    # type in nodes.. nodes[node_name]['function']
+
+
+
+
+
+
+    
+
+
+    # Temporary:    
+    # print('props: ', dir(node))
+    last_child_end = node.last_child().end_col_offset
```

Comment:
This is indeed the approach I suggested. You should remove the "as_string" stuff above.

Use this approach for setting the line number as well.